pkg/ruler/rulestore/bucket_client.go

```diff
@@ -0,0 +1,312 @@
+package rulestore
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/cortexproject/cortex/pkg/ruler/rules"
+	"github.com/cortexproject/cortex/pkg/storage/bucket"
+	"github.com/cortexproject/cortex/pkg/tenant"
+)
+
+const (
+	delim      = "/"
+	rulePrefix = "rules"
+)
+
+type BucketRuleStore struct {
+	bucket          objstore.Bucket
+	loadConcurrency int
+
+	logger log.Logger
+}
+
+func NewBucketRuleStore(bucket objstore.Bucket, loadConcurrency int, logger log.Logger) *BucketRuleStore {
+	return &BucketRuleStore{
+		bucket:          bucket,
+		loadConcurrency: loadConcurrency,
+		logger:          logger,
+	}
+}
+
+func newRuleStoreUserBucketClient(userID string, bkt objstore.Bucket) objstore.Bucket {
+	return bucket.NewUserBucketClient(userID, bucket.NewPrefixedBucketClient(bkt, rulePrefix))
```

Comment:
I'd suggest adding a test to verify that user-bucket and prefixed-bucket are composed correctly, and produce expected paths (rules/user1/...).

Reply:
I ended up removing this function since it's superflous when the prefixed bucket is a field on the rulestore. In terms of a test for this, wouldn't it be enough to pass the existing rulestore tests to verify the user/prefix bucket are composed correctly?

Reply:
> In terms of a test for this, wouldn't it be enough to pass the existing rulestore tests to verify the user/prefix bucket are composed correctly?

You are right. I can see that tests already check for full path of some objects.

Would it make sense to refactor tests and reuse them for both bucket-based rulestore and objectclient-based one?

Reply:
I think that would make sense. However, I may update the tests to use the rule store `Get` method instead of directly checking the underlying bucket if an object has been deleted. This will make it easier to write a reusable test having to consider the underlying implementation.

Reply:
I agree, I'm hesitant to do it in this PR since it will push the LOC changes in this PR beyond where it's already at. Is it ok if I do this in a separate PR?

Reply:
Sounds good to me.